source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -1253,6 +1414,51 @@ void Filter::disableRetryTimer() {
   }
 }
 
+void Filter::checkUpstreamConnectionTrigger() {
+  // Check if we should establish the upstream connection based on current state.
+  using Mode = envoy::extensions::filters::network::tcp_proxy::v3::UpstreamConnectMode;
+
+  bool should_connect = false;
+
+  switch (connect_mode_) {
+  case Mode::IMMEDIATE:
+    // Should have already connected.
+    return;
+
+  case Mode::ON_DOWNSTREAM_DATA:
+    should_connect = initial_data_received_;
+    ASSERT(should_connect); // If we're calling this for ON_DOWNSTREAM_DATA, initial_data_received_
+                            // must be true.
+    break;
+
+  case Mode::ON_DOWNSTREAM_TLS_HANDSHAKE:
+    should_connect = tls_handshake_complete_ || !waiting_for_tls_handshake_;
+    break;
+
+  default:
+    // Handle sentinel values.
+    ENVOY_BUG(false, "Invalid connection mode");
+    return;
+  }
+
+  ASSERT(should_connect); // We should only call this when we want to connect.
+  ASSERT(!connecting_);   // We shouldn't be connecting already.
+  ASSERT(!upstream_);     // We shouldn't have an upstream already.
+
+  // Route should already be set in onNewConnection(), but set it again if not.
+  if (route_ == nullptr) {
+    route_ = pickRoute();
+  }
+  ASSERT(route_ != nullptr);
+  establishUpstreamConnection();
+}
```

Comment:
Is this still make sense because I think almost every case is handled in other code logic?